pkg/vsphere/toolbox/process.go

```diff
@@ -0,0 +1,359 @@
+	}
+}
+
+// ProcessState is the toolbox representation of the GuestProcessInfo type
+type ProcessState struct {
+	Name      string
+	Args      string
+	Owner     string
+	Pid       int64
+	ExitCode  int32
+	StartTime int64
+	EndTime   int64
+}
+
+func (s *ProcessState) toXML() string {
+	const format = "<proc>" +
+		"<cmd>%s</cmd>" +
+		"<name>%s</name>" +
+		"<pid>%d</pid>" +
+		"<user>%s</user>" +
+		"<start>%d</start>" +
+		"<eCode>%d</eCode>" +
+		"<eTime>%d</eTime>" +
+		"</proc>"
+
+	name := filepath.Base(s.Name)
+
+	argv := []string{s.Name}
+
+	if len(s.Args) != 0 {
+		argv = append(argv, xmlEscape.Replace(s.Args))
+	}
+
+	args := strings.Join(argv, " ")
+
+	exit := atomic.LoadInt32(&s.ExitCode)
+	end := atomic.LoadInt64(&s.EndTime)
+
+	return fmt.Sprintf(format, name, args, s.Pid, s.Owner, s.StartTime, exit, end)
+}
+
+// Process managed by the ProcessManager.
+type Process struct {
+	ProcessState
+
+	Start func(context.Context, *VixMsgStartProgramRequest) (int64, error)
+	Wait  func() error
+	Kill  context.CancelFunc
+
+	ctx context.Context
+}
+
+// ProcessError can be returned by the Process.Wait function to propagate ExitCode to ProcessState.
+type ProcessError struct {
+	Err      error
+	ExitCode int32
+}
+
+func (e *ProcessError) Error() string {
+	return e.Err.Error()
+}
+
+// ProcessManager manages processes within the guest.
+// See: http://pubs.vmware.com/vsphere-60/topic/com.vmware.wssdk.apiref.doc/vim.vm.guest.ProcessManager.html
+type ProcessManager struct {
+	wg      sync.WaitGroup
+	mu      sync.Mutex
+	expire  time.Duration
+	entries map[int64]*Process
+	pids    sync.Pool
+}
+
+// NewProcessManager creates a new ProcessManager instance.
+func NewProcessManager() *ProcessManager {
+	pid := int64(32768) // TODO: /proc/sys/kernel/pid_max
+
+	return &ProcessManager{
+		expire:  time.Minute * 5,
+		entries: make(map[int64]*Process),
+		pids: sync.Pool{
+			New: func() interface{} {
+				return atomic.AddInt64(&pid, 1)
+			},
+		},
+	}
+}
+
+// Start calls the Process.Start function, returning the pid on success or an error.
+// A goroutine is started that calls the Process.Wait function.  After Process.Wait has
+// returned, the ProcessState EndTime and ExitCode fields are set.  The process state can be
+// queried via ListProcessesInGuest until it is removed, 5 minutes after Wait returns.
+func (m *ProcessManager) Start(r *VixMsgStartProgramRequest, p *Process) (int64, error) {
+	p.Name = r.ProgramPath
+	p.Args = r.Arguments
+
+	// Owner is cosmetic, but useful for example with: govc guest.ps -U $uid
+	if p.Owner == "" {
+		p.Owner = defaultOwner
+	}
+
+	p.StartTime = time.Now().Unix()
+
+	p.ctx, p.Kill = context.WithCancel(context.Background())
+
+	pid, err := p.Start(p.ctx, r)
+	if err != nil {
+		return -1, err
+	}
+
+	if pid == 0 {
+		p.Pid = m.pids.Get().(int64) // pseudo pid for funcs
+	} else {
+		p.Pid = pid
+	}
+
+	m.mu.Lock()
+	m.entries[p.Pid] = p
+	m.mu.Unlock()
+
+	m.wg.Add(1)
+	go func() {
+		werr := p.Wait()
```

Comment:
Faiyaz had a similar problem before, I believe he wanted to run a command and collect its exitcode but had issues because of the child reaper in the tether. I'm wondering whether this will race with it?